Win32 merge

pheest opened the pull request:
In Makehelper.py, replaced use of Linux-specific library path configuration variables with get_python_lib() call.

In Rules.py, added use of xcopy to copy Python files to the install path for Windows use.

In devsupapp/source/dbfield.c, added support capability for lsi record type by:
    replacing fixed 40-charachter string limit with variable length.
    Calling the 'special' method of lsiRecord.c to set the record string length.

In devsupapp/source/makefile:
    A Windows Python loadable module has the file extenstion '.pyd' (not .dll). Overrode the SHRLIB_SUFFIX_BASE variable to apply this.
    Added new files _dbapi.dbd, _int64.dbd and _lsilso.dbd. Set these to copy to the same folder as Python files (although they aren't).

In devsupapp/source/devsup:
    added _dbapi.dbd, _int64.dbd and _lsilso.dbd
    in __init__.py, replaced use of temporary file with the above static files. _int64 and _lsilso are loaded conditionally on the EPICS base version.

In devsupapp/source/devsup/disect.py:
    InstanceType is imported at line 6 for Python 2 usage, and will raise an exception with Python 3.
    This is tested at line 84 with elif InstanceType is not None.
    But when the excpetion is raised, InstanceType is not None, it is undefined - causing the application to fail.
    Added InstanceType = None to ensure this doesn't happen.

In devsuppapp/source added changes to support alarm handling on output PVs, according to the mail sent on 06/07.
To control the stat and sevr record fields.

In pyiocapp/setup.c added the definition PATH_MAX (as _MAX_PATH) for Windows use.
In devsupapp/source/devsup/__init__.py _init() EPICS base.dbd is loaded on condition 'if not iocMain'.
But In pyiocapp/setup.c passed 'iocMain = True' in pySetupReg() ... so where is EPICS base.dbd loaded in _dbapi?
Changed to iocMain = False.

pheest committed:
Partially working on Windows.

pheest committed:
Re-tested on Windows and Linux.

pheest committed:
Corrected _dbapi VS project name.

pheest committed:
Added report pyDevSup.docx

pheest committed:
Follows document review by kwoods.

pheest committed:
Updated to VS2022 and Python 3.11

pheest committed:
Got it working with Michael's help.

pheest committed:
Deleted report document.

pheest committed:
Was able to run from makefiles in cmd.

pheest committed:
Built & ran hidenrga project.

pheest committed:
Makefile build and test on Windows.

pheest committed:
* The construct @[ -d $(dir $@) ] and install command fail on Windows, resulting in build errors
Replaced with use of xcopy for windows - a better approach could be found, I think.

A DLL imported as a Python module needs to have the extension .pyd - overrode SHRLIB_SUFFIX_BASE to create _dbapi.pyd.

Also added Windows library dependencies.

pheest committed:
Compared and merged with later changes made by mdavidsaver.

pheest committed:
Moved tempoary database file in to permanent file _dbapi.dbd.

pheest committed:
Separated int64 and lsi/lso from the main _dbapi.dbd file in order to make them conditional on epics version number.

pheest committed:
Removed unnecessary debug print statements.

pheest committed:
Corrected address operator mistake.

pheest committed:
For the purpose of making a pull request, discarded changes that I want to keep in main branch.

mdavidsaver commented on devsupApp/src/Makefile:
```diff
@@ -9,6 +9,9 @@ include $(TOP)/configure/CONFIG_PY
 
 LOADABLE_LIBRARY_HOST += _dbapi
+ifeq ($(OS),Windows_NT)
+    SHRLIB_SUFFIX_BASE = .pyd
```
I think this should be in `configure/CONFIG_PY`, where the equivalent for Darwin/OSX is set.

```make
ifeq ($(OS_CLASS),WIN32)
LOADABLE_SHRLIB_SUFFIX = .pyd
endif
```

pheest replied:
OK, but I get a build error - which I can resolve by setting:
LOADABLE_LIBRARY_HOST_DEFAULT += _dbapi
LOADABLE_LIBRARY_HOST_WIN32 += _dbapi.pyd

pheest replied:
It does need to be SHRLIB_SUFFIX_BASE (setting LOADABLE_SHRLIB_SUFFIX does not have the desired effect).
But this can be, as you suggest, in CONFIG.PY.

mdavidsaver commented on devsupApp/src/devsup/__init__.py:
```diff
@@ -45,41 +45,28 @@
+    _dbapi.dbReadDatabase(dbd_name)
+    # e.g. 3.15.0.2 -> 30000 + 1500 + 0 + 2 = 31502
+    epics_version_int = EPICS_VERSION * 10000 + EPICS_REVISION * 100 + EPICS_MODIFICATION * 10 + EPICS_PATCH_LEVEL
+    if epics_version_int >= 31502:
```
While this will certainly work,  fyi. tuple comparison is shorter and easier to understand.  (cf. `sys.version_info`)

```py
epics_version_int = (EPICS_VERSION, EPICS_REVISION, EPICS_MODIFICATION, EPICS_PATCH_LEVEL)
if epics_version_int >= (3, 15, 0, 2):
```

pheest replied:
Noted.

mdavidsaver commented on makehelper.py:
```diff
@@ -14,16 +14,26 @@
+    from distutils.sysconfig import get_python_inc, get_python_lib
+except ImportError:
+    def get_python_inc():
+        return get_config_var('INCLUDEPY') or ''
```
Is a fallback `get_python_lib()` needed here as well?

mdavidsaver commented on pyIocApp/setup.c:
```diff
@@ -23,6 +23,10 @@
 #include "pydevsup.h"
+#ifdef _WIN32
+#define PATH_MAX _MAX_PATH
+#include <direct.h>
```
Does the order of these two lines change the behavior of `direct.h`?

pheest replied:
No, I don't think so. It's more appropriate the other way round.
And _MAX_PATH is actually defined in stdlib.h.

mdavidsaver commented on pyIocApp/setup.c:
```diff
@@ -135,7 +139,7 @@ static void pySetupReg(void)
 
     if(PyRun_SimpleString("import devsup\n"
-                          "devsup._init(iocMain=True)\n"
+                          "devsup._init(iocMain=False)\n"
```
Is this change necessary now that `_init()` does not use `NamedTemporaryFile`?

pheest replied:
The two are separate in _init(). I'm not too sure what was intended here, but I do not see any existing call of _init() that caused base to load, and this is still the case.

pheest replied:
base would of course have been loaded by the IOC start-up.
So I'm not sure whether loading it from _dbapi is actually required anyway.

mdavidsaver commented on requirements-deb10.txt:
```diff
@@ -1,2 +1,3 @@
 numpy==1.16.2
-nose==1.3.7
+#nose==1.3.7
```
I think references to nose v1 can be entirely removed.

mdavidsaver reviewed (commented):
Looks good aside from some minor cleanup.

mdavidsaver commented:
Some files which I don't think should be changed/added by this PR.

- configure/RELEASE
- configure/RELEASE.windows-x64.common
- pyDevSup.sln

mdavidsaver commented:
I am not sure why the github actions jobs have not started.  Maybe the conflict in `makehelper.py`.  Maybe something to do with an ongoing issue.  https://www.githubstatus.com/

pheest committed:
Updated following review comments by @mdavidsaver.

pheest committed:
Merge branch 'win32' into win32-merge

pheest committed:
Updated RULES_PY to use @$(INSTALL). This requires include $(CONFIG)/CONFIG_BASE for the definition of the INSTALL macro.

epicsThreadExitMain is deprecated by recent EPICS base versions (and resolves to cantProceed). Being unsure as to what best to do about this, I commented it out in devSupMain.cpp.

Python 2.7 has been obsoleted and won't install on the runner without taking special measures.
Being unsure what to do about this, I have commented out builds that invoke 2.7.

Github Windows-latest (i.e. Windows server 2022) is installed with VS2022 (not VS2019).
Use of the debug build configuration will seek to link with the Python debug libraries that are not normally installed. I have a solution to this problem but we should debate how to proceed.

Github Windows server 2019 is, I think, installed with VS2019, this should build OK with default configuration.

pheest committed:
Merge branch 'win32-merge' of https://github.com/pheest/pyDevSup into win32-merge

mdavidsaver commented:
> Please may I query the intended purpose of the PY_LDLIBS variable that is output from makehelper.py?

```py
from sysconfig import get_config_var
print(get_config_var('BLDLIBRARY'))
```

On linux/debian 12 this prints `-lpython3.11`.

In *nix convention, `LDLIBS` contains `-l*` entries mentioning library names, while `LDFLAGS` has `-L*` setting the search path.

In this context, my goal was that `PY_LDLIBS` contain `-lpython*` as well as any additional dependencies.  Possibly libraries like `-ldl`, `-lz`, or maybe winsock.

pheest committed:
Added use of the PY_LDLIBS variable to locate the Python import library.
On Windows, the library isn't in a system folder so the fully-qualified path is required.

mdavidsaver commented:
(I am not sure why GHA is making me approve each run?)

Sorry, I had forgotten how few of the sysconfig variables appear in windows builds...  which will likely need to remain a special case.

fyi. `python -m sysconfig` on Windows shows.

```
 	BINDIR = "C:\hostedtoolcache\windows\Python\3.10.11\x64"
	BINLIBDEST = "C:\hostedtoolcache\windows\Python\3.10.11\x64\Lib"
	EXE = ".exe"
	EXT_SUFFIX = ".cp310-win_amd64.pyd"
	INCLUDEPY = "C:\hostedtoolcache\windows\Python\3.10.11\x64\Include"
	LIBDEST = "C:\hostedtoolcache\windows\Python\3.10.11\x64\Lib"
	SO = ".cp310-win_amd64.pyd"
	TZPATH = ""
	VERSION = "310"
	abiflags = ""
	base = "C:\hostedtoolcache\windows\Python\3.10.11\x64"
	exec_prefix = "C:\hostedtoolcache\windows\Python\3.10.11\x64"
	installed_base = "C:\hostedtoolcache\windows\Python\3.10.11\x64"
	installed_platbase = "C:\hostedtoolcache\windows\Python\3.10.11\x64"
	platbase = "C:\hostedtoolcache\windows\Python\3.10.11\x64"
	platlibdir = "lib"
	prefix = "C:\hostedtoolcache\windows\Python\3.10.11\x64"
	projectbase = "C:\hostedtoolcache\windows\Python\3.10.11\x64"
	py_version = "3.10.11"
	py_version_nodot = "310"
	py_version_nodot_plat = "310"
	py_version_short = "3.10"
	srcdir = "C:\hostedtoolcache\windows\Python\3.10.11\x64"
	userbase = "C:\Users\runneradmin\AppData\Roaming\Python"
```

pheest commented:
So the order of libraries passed to the linker matters. See https://stackoverflow.com/questions/11893996/why-does-the-order-of-l-option-in-gcc-matter. LDFLAGS go before LIBS and o files. Passing the library by means of LDFLAGS may not be such a good idea, after all.

Building with debug configuration means dependency on on the debug python library (python310_d.lib), which isn't installed.

There are some specific instructions regarding macOS at https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md.

pheest committed:
Reverted change to apply the Python library through LDFLAGS only due to the linker order issue.

Commissioned use of nose testing on Windows.
This requires:
a. (already noted) issue regarding simultaneous temporary file read and write to be addressed in util.py.
b. Windows .pyd file does not search PATH for DLLs - instead the EPICS base path has to be explicitly added in __init__.py with os.add_dll_directory.

Tested with Python 3.11 on Windows.

pheest committed:
Reverted change to use get_python_lib - it seems not to work correctly for earlier Python versions.

pheest committed:
Does it need __init__.py?

pheest committed:
__init__.py needed in build directory?

pheest committed:
Removed @ from Windows nose test command - so I can see what command is being passed.
Reverted unnecessary changes to __init__.py usage.

pheest committed:
Additional debugging added to check directory paths.

pheest committed:
/c/hostedtoolcache/ syntax?

pheest committed:
Oops, always the slashes that get you.

pheest committed:
Make sure we use any existing PYTHONPATH.

pheest committed:
Silly syntax mistake.

pheest committed:
Silly, PYTHONPATH doesn't exist until we define it.

pheest committed:
Try pushd / popd instead of set PYTHONPATH.

pheest committed:
Removed debugging statements - these are not helping.

pheest committed:
Updated CI-scripts as advised by mdavidsaver.

pheest committed:
Put debug messages back in __init__.py and replaced slashes to use Windows style.

pheest committed:
Removed use of XEPICS_BASE env variable as EPICS_BASE is already an env variable.

pheest committed:
Removed debug print statements again.

pheest committed:
Added tests & logic for lsi, lso and int64 types.

ralphlange commented:
@mdavidsaver: Where are we on this?
To me, it looks as if there are only minor issues remaining. (If any.)

Background: Peter uses this on ITER-related work, so I need to include it in our CODAC packaging of pyDevSup. Obviously, I would prefer an upstream merge, as opposed to something I do only for our CODAC packaging = that I might need to clean up later...

eddybl commented:
There will be a conflict in `makehelper.py` if #39  is merged

pheest committed:
Resolved merge conflict of makehelper.py

pheest committed:
Addressed numpy 2.0 build error - see https://github.com/pytorch/pytorch/issues/121798

pheest committed:
Corrected comment typo.

pheest committed:
Changed to use master branch of .ci

pheest commented:
The conflict has been addressed.
This also required to address the numpy 2.0 issue noted by another user.

eddybl commented:
I actually do not think your fix to `makehelper.py` will work, as `distutils.sysconfig` and `sysconfig` are not replaceable and `get_python_inc` is not available from `sysconfig`

But maybe we should just wait for the other PR to be merged (or maybe modified). I just wanted to note down that there might be a potential change required here down the line, depending on the PR resolve order

pheest commented:
It is working. It is set up to build on GitHub with a selection of
platforms and Python versions; these show green.
I understand that distutils is deprecated, but I hadn't been aware of your
second point.
The exception block does appropriately implement get_python_inc if this
isn't available.


On Fri, Aug 9, 2024 at 12:56 PM eddybl ***@***.***> wrote:

> I actually do not think your fix to makehelper.py will work, as
> distutils.sysconfig and sysconfig are not replaceable and get_python_inc
> is not available from sysconfig
>
> But maybe we should just wait for the other PR to be merged (or maybe
> modified). I just wanted to note down that there might be a potential
> change required here down the line, depending on the PR resolve order
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mdavidsaver/pyDevSup/pull/33#issuecomment-2277779762>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD6SZN3UOLWHOUK3TUAHJTTZQSU5JAVCNFSM6AAAAABLOTLSZKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDENZXG43TSNZWGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

pheest committed:
Reverted undesired change of distutils.sysconfig

pheest commented:
I confirm that my version does not work with Python 3.12, due to distutils being removed.

@eddybl, I saw your pull request https://github.com/mdavidsaver/pyDevSup/pull/39 .
That is considerably cleaner, thank you.

I confirm that it works with Linux to 3.12.

It does not work (as it stands) with Windows because:

	The library directory is not contained within 'LIBDIR' but within 'LIBDEST'.
	
	'BASECFLAGS' does not exist, so the compiler tries to build a non-existent file called 'Null'.
	
	get_python_version() returns e.g. 3.11 (both Linux and Windows), so the linker tries to use python3.11.lib - but the file is actually python311.lib.
	'VERSION' contains '3.11' on Linux but '311' on Windows, so that works for both.
	
	The Windows Python library path is not a system path, so that needs to be explicitly specified.

I have created a merged version from yours, and tested it in my 'devel' branch, with the necessary changes for Windows use.
I have verified that it works for all Python versions to 3.12 on both Linux and Windows.

I have not added this to my pull request, to date.
Should I?

eddybl commented:
Yes, so i have not considered windows during my changes. 

I think it makes sense to push your changes from devel also to this PR.

How the PRs will be resolved is then a question for @mdavidsaver: For example if this PR might get merged soonish, there is no need to merge #39. But if this PR might still be open for a while it might make sense to already merge #39 now, because it is a rather small fix and in that case we most likely have to resolve a small merge conflict for this PR, but that should be doable.

pheest committed:
Updated makehelper.py with support for Windows using the file version of #39.

pheest committed:
Also updated the Github build configuration which now includes Python 3.12.

mdavidsaver commented:
> How the PRs will be resolved is then a question for @mdavidsaver ...

At this point, I think it is apparent to everyone that I don't have sufficient time to act as maintainer of pyDevSup, and it isn't clear when I will.  I also don't want my lack of time to be a barrier to other who are willing to contribute.

So what to do?

Are others willing to take over review?  Full maintainer?  Transfer this repository?  A pyDevSup2 fork?  (in keeping with my unimaginative naming...)

pheest commented:
Thank you @mdavidsaver for being so open.

I am willing to be a maintainer.

My thoughts:
	I think there should ideally be more than one maintainer.
	I take the view that the module should be hosted in an organizational site, rather than an individual person's.
	I believe maintainers should not approve their own pull requests.
	I don't see a need to change the module name.

pheest committed:
Updated at end of project.

pheest committed:
Update ci-scripts-build.yml to allow manual workflow

Commit:
Fixed problem with NPY_CARRAY and NPY_CARRAY_RO being unavailable in newer versions of numpy.

Commit:
Windows server 2019 no longer available.

Commit:
Tested Windows (and Linux) builds with Base 3.15.
Excluded int64 and long string for that reason.
Base 3.14 doesn't work because RuntimeError: Requires Base >=3.15.

Commit:
Oops, corrected mistakes.

Commit:
Moved toward tford's branch with some small changes.

Commit:
Re-engineered the alarm handling stuff.
Updated the readme file with changed repository location.
